core/server/common/src/main/java/alluxio/master/journal/raft/SnapshotReplicationManager.java

```diff
@@ -0,0 +1,535 @@
+
+    /** A snapshot is being installed to the journal storage. */
+    INSTALLING,
+  }
+
+  private final AtomicReference<DownloadState> mDownloadState =
+      new AtomicReference<>(DownloadState.IDLE);
+
+  /**
+   * @param journalSystem the raft journal system
+   * @param storage the snapshot storage
+   */
+  public SnapshotReplicationManager(RaftJournalSystem journalSystem,
+      SimpleStateMachineStorage storage) {
+    mStorage = storage;
+    mJournalSystem = journalSystem;
+  }
+
+  /**
+   * @param journalSystem the raft journal system
+   * @param storage the snapshot storage
+   */
+  @VisibleForTesting
+  SnapshotReplicationManager(RaftJournalSystem journalSystem,
+      SimpleStateMachineStorage storage, RaftJournalServiceClient client) {
+    this(journalSystem, storage);
+    mJournalServiceClient = client;
+  }
+
+  /**
+   * Downloads and installs a snapshot from the leader.
+   *
+   * @return a future with the term index of the installed snapshot
+   */
+  public CompletableFuture<TermIndex> installSnapshotFromLeader() {
+    if (mJournalSystem.isLeader()) {
+      return RaftJournalUtils.completeExceptionally(
+          new IllegalStateException("Abort snapshot installation after becoming a leader"));
+    }
+    if (!transitionState(DownloadState.IDLE, DownloadState.STREAM_DATA)) {
+      return RaftJournalUtils.completeExceptionally(
+          new IllegalStateException("State is not IDLE when starting a snapshot installation"));
+    }
+    try {
+      RaftJournalServiceClient client = getJournalServiceClient();
+      String address = String.valueOf(client.getAddress());
+      SnapshotDownloader<DownloadSnapshotPRequest, DownloadSnapshotPResponse> observer =
+          SnapshotDownloader.forFollower(mStorage, address);
+      client.downloadSnapshot(observer);
+      return observer.getFuture().thenApplyAsync((termIndex) -> {
+        mDownloadedSnapshot = observer.getSnapshotToInstall();
+        transitionState(DownloadState.STREAM_DATA, DownloadState.DOWNLOADED);
+        long index = installDownloadedSnapshot();
+        if (index == RaftLog.INVALID_LOG_INDEX) {
+          throw new CompletionException(new RuntimeException(
+              String.format("Failed to install the downloaded snapshot %s", termIndex)));
+        }
+        if (index != termIndex.getIndex()) {
+          throw new CompletionException(new IllegalStateException(
+              String.format("Mismatched snapshot installed - downloaded %d, installed %d",
+                  termIndex.getIndex(), index)));
+        }
+        return termIndex;
+      }).whenComplete((termIndex, throwable) -> {
+        if (throwable != null) {
+          LOG.error("Unexpected exception downloading snapshot from leader {}.", address,
+              throwable);
+          transitionState(DownloadState.STREAM_DATA, DownloadState.IDLE);
+        }
+      });
+    } catch (Exception e) {
+      transitionState(DownloadState.STREAM_DATA, DownloadState.IDLE);
```

Comment:
Is there any chance the current state is NOT `STREAM_DATA`? Then, we would not be able to reset the state, right?

Reply:
Please see the comment above.